rust/src/types/observer.rs

```diff
@@ -0,0 +1,52 @@
+use std::borrow::Cow;
+
+use crate::interpreter::Interpreter;
+
+pub trait Observer {
+    fn pre_op(&mut self, interpreter: &Interpreter);
+
+    fn post_op(&mut self, interpreter: &Interpreter);
+
+    fn log(&mut self, message: Cow<str>);
+}
+
+pub struct NoOpObserver();
+
+impl Observer for NoOpObserver {
+    fn pre_op(&mut self, _interpreter: &Interpreter) {}
+
+    fn post_op(&mut self, _interpreter: &Interpreter) {}
+
+    fn log(&mut self, _message: Cow<str>) {}
+}
+
+pub struct LoggingObserver();
+
+impl Observer for LoggingObserver {
+    fn pre_op(&mut self, interpreter: &Interpreter) {
+        println!(
+            "pre opcode={:?} gas={} stack-size={}",
```

Comment:
Is this compatible with the printed logs of lfvm? is the format equal?

Reply:
it was not, but I adjusted it now.